doc/man/man3/MainComponent.3
.TH "MainComponent" 3Version 1.0.1" "CxxWin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MainComponent \- The '\fBMainComponent\fP' class\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <MainComponent\&.hpp>\fP
.PP
Inherits \fBMainWindow\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBMainComponent\fP ()"
.br
.RI "Construct a new '\fBMainComponent\fP' object\&. "
.ti -1c
.RI "PCWSTR \fBClassName\fP () CONST"
.br
.ti -1c
.RI "LRESULT \fBHandleMessage\fP (UINT uMsg, WPARAM wParam, LPARAM lParam)"
.br
.RI "A typical window procedure is simply a large 'switch()' statement that switches on the 'uMsg' code\&. "
.ti -1c
.RI "void \fBCalculateLayout\fP ()"
.br
.ti -1c
.RI "HRESULT \fBCreateGraphicsResources\fP ()"
.br
.RI "Initialize graphics resources such as brushes, colours, and layouts for further processing\&. "
.ti -1c
.RI "void \fBDiscardGraphicsResources\fP ()"
.br
.ti -1c
.RI "HRESULT \fBOnCreate\fP ()"
.br
.ti -1c
.RI "void \fBOnPaint\fP ()"
.br
.ti -1c
.RI "void \fBResize\fP ()"
.br
.ti -1c
.RI "void \fBOnClose\fP () CONST"
.br
.ti -1c
.RI "void \fBOnDestroy\fP ()"
.br
.in -1c

Public Member Functions inherited from \fBMainWindow\fP
.in +1c
.ti -1c
.RI "\fBMainWindow\fP ()=default"
.br
.RI "Default-Construct a new Main Window object\&. "
.ti -1c
.RI "PCWSTR \fBClassName\fP () CONST"
.br
.ti -1c
.RI "LRESULT \fBHandleMessage\fP (UINT uMsg, WPARAM wParam, LPARAM lParam)"
.br
.in -1c

Public Member Functions inherited from \fBBaseWindow< MainWindow >\fP
.in +1c
.ti -1c
.RI "\fBBaseWindow\fP ()"
.br
.RI "Construct a new '\fBBaseWindow\fP' object\&. "
.ti -1c
.RI "BOOL \fBCreate\fP (PCWSTR lpWindowName, DWORD dwStyle, DWORD dwExStyle=0, int x=CW_USEDEFAULT, int y=CW_USEDEFAULT, int nWidth=CW_USEDEFAULT, int nHeight=CW_USEDEFAULT, HWND hWndParent=0, HMENU hMenu=0)"
.br
.RI "Create a new '\fBBaseWindow\fP' -derived object\&. "
.ti -1c
.RI "HWND \fBWindow\fP () CONST"
.br
.RI "Return the handle to the '\fBBaseWindow\fP' -derived object\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "ID2D1Factory * \fBpFactory\fP"
.br
.ti -1c
.RI "ID2D1HwndRenderTarget * \fBpRenderTarget\fP"
.br
.ti -1c
.RI "ID2D1SolidColorBrush * \fBpBrush\fP"
.br
.ti -1c
.RI "ID2D1SolidColorBrush * \fBpBrushB\fP"
.br
.ti -1c
.RI "ID2D1SolidColorBrush * \fBpBrushC\fP"
.br
.ti -1c
.RI "D2D1_ELLIPSE \fBellipseA\fP"
.br
.ti -1c
.RI "D2D1_ELLIPSE \fBellipseB\fP"
.br
.ti -1c
.RI "D2D1_ELLIPSE \fBellipseC\fP"
.br
.ti -1c
.RI "D2D1_POINT_2F \fBptMouse\fP"
.br
.in -1c
.SS "Additional Inherited Members"


Static Public Member Functions inherited from \fBBaseWindow< MainWindow >\fP
.in +1c
.ti -1c
.RI "static LRESULT CALLBACK \fBWindowProc\fP (HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)"
.br
.RI "'LRESULT' is an integer value that your program returns to Windows\&. "
.in -1c
.in +1c
.ti -1c
.RI "virtual PCWSTR \fBClassName\fP () CONST PURE"
.br
.RI "Return a string of the name of '\fBBaseWindow\fP' -derived object\&. "
.ti -1c
.RI "virtual LRESULT \fBHandleMessage\fP (UINT uMsg, WPARAM wParam, LPARAM lParam) PURE"
.br
.RI "Handle the incoming messages\&. "
.in -1c

Protected Attributes inherited from \fBBaseWindow< MainWindow >\fP
.in +1c
.ti -1c
.RI "HWND \fBm_hwnd\fP"
.br
.RI "The handle to the '\fBBaseWindow\fP' -derived object\&. "
.in -1c
.SH "Detailed Description"
.PP 
The '\fBMainComponent\fP' class\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "MainComponent::MainComponent ()"

.PP
Construct a new '\fBMainComponent\fP' object\&. 
.SH "Member Function Documentation"
.PP 
.SS "void MainComponent::CalculateLayout ()"

.SS "PCWSTR MainComponent::ClassName ()\fC [virtual]\fP"

.PP
\fBReturns\fP
.RS 4
PCWSTR 
.RE
.PP

.PP
Reimplemented from \fBBaseWindow< MainWindow >\fP\&.
.SS "HRESULT MainComponent::CreateGraphicsResources ()"

.PP
Initialize graphics resources such as brushes, colours, and layouts for further processing\&. 
.PP
\fBReturns\fP
.RS 4
HRESULT 
.RE
.PP

.SS "void MainComponent::DiscardGraphicsResources ()"

.SS "LRESULT MainComponent::HandleMessage (UINT uMsg, WPARAM wParam, LPARAM lParam)\fC [virtual]\fP"

.PP
A typical window procedure is simply a large 'switch()' statement that switches on the 'uMsg' code\&. Add cases for each message that you want to handle\&.
.PP
\fBParameters\fP
.RS 4
\fIuMsg\fP 
.br
\fIwParam\fP 
.br
\fIlParam\fP 
.RE
.PP
\fBReturns\fP
.RS 4
LRESULT 
.RE
.PP

.PP
Reimplemented from \fBBaseWindow< MainWindow >\fP\&.
.SS "void MainComponent::OnClose ()"

.SS "HRESULT MainComponent::OnCreate ()"

.PP
\fBReturns\fP
.RS 4
HRESULT 
.RE
.PP

.SS "void MainComponent::OnDestroy ()"

.SS "void MainComponent::OnPaint ()"

.SS "void MainComponent::Resize ()"

.SH "Member Data Documentation"
.PP 
.SS "D2D1_ELLIPSE MainComponent::ellipseA\fC [private]\fP"

.SS "D2D1_ELLIPSE MainComponent::ellipseB\fC [private]\fP"

.SS "D2D1_ELLIPSE MainComponent::ellipseC\fC [private]\fP"

.SS "ID2D1SolidColorBrush* MainComponent::pBrush\fC [private]\fP"

.SS "ID2D1SolidColorBrush* MainComponent::pBrushB\fC [private]\fP"

.SS "ID2D1SolidColorBrush* MainComponent::pBrushC\fC [private]\fP"

.SS "ID2D1Factory* MainComponent::pFactory\fC [private]\fP"

.SS "ID2D1HwndRenderTarget* MainComponent::pRenderTarget\fC [private]\fP"

.SS "D2D1_POINT_2F MainComponent::ptMouse\fC [private]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for CxxWin from the source code\&.


include/CxxWin/MainComponent.cpp
/***************************************************************************//**
 * @file MainComponent.cpp
 * @author StoneyDSP (nathanjhood@googlemail.com)
 * @brief Contains the definitions of 'MainComponent'.
 * @version 1.0.0
 * @date 2023-08-28
 *
 * @copyright Copyright (c) 2023
 *
 ******************************************************************************/

// #include "modules/Win32_core/win32_core.mm"

#include "MainComponent.hpp"

/**
 * @brief Construct a new 'MainComponent' object.
 *
 */
MainComponent::MainComponent()
  : pFactory (NULL)
  , pRenderTarget (NULL)
  , pBrush (NULL)
  , pBrushB (NULL)
  , pBrushC (NULL)
  , ellipseA (D2D1::Ellipse(D2D1::Point2F(), 0.0F, 0.0F))
  , ellipseB (D2D1::Ellipse(D2D1::Point2F(), 0.0F, 0.0F))
  , ellipseC (D2D1::Ellipse(D2D1::Point2F(), 0.0F, 0.0F))
  , ptMouse (D2D1::Point2F())
{

}

/**
 * @brief
 *
 * @return PCWSTR
 */
PCWSTR MainComponent::ClassName() const
{
    return CXXWIN_LONG_NAME;
}

/**
 * @brief A typical window procedure is simply a large 'switch()' statement that
 * switches on the 'uMsg' code. Add cases for each message that you want to
 * handle.
 *
 * @param uMsg
 * @param wParam
 * @param lParam
 * @return LRESULT
 */
LRESULT MainComponent::HandleMessage(UINT uMsg, WPARAM wParam, LPARAM lParam)
{
    switch (uMsg)
    {
    case WM_CREATE:

        if (FAILED(OnCreate()))
        {
            return -1;
        }
        DPIScale::Initialize(pFactory);
        return 0;

    case WM_CLOSE:

        OnClose();
        return 0;

    case WM_DESTROY:

        OnDestroy();
        return 0;

    case WM_SIZE:
        /**
        {
            int width = LOWORD(lParam);
            int height = HIWORD(lParam);

            OnSize(m_hwnd, (UINT)wParam, width, height);
        }
        */
        Resize();
        return 0;

    case WM_PAINT:
        OnPaint();
        return 0;

    default:

        return DefWindowProc(m_hwnd, uMsg, wParam, lParam);
    }

    return TRUE;
}

/**
 * @brief
 *
 */
void MainComponent::Resize()
{
    if (pRenderTarget != NULL)
    {
        RECT rc;
        GetClientRect(m_hwnd, &rc);

        D2D1_SIZE_U size = D2D1::SizeU(rc.right, rc.bottom);

        pRenderTarget->Resize(size);
        CalculateLayout();
        InvalidateRect(m_hwnd, NULL, FALSE);
    }
}

/**
 * @brief
 *
 */
void MainComponent::CalculateLayout()
{
    if (pRenderTarget != NULL)
    {
        D2D1_SIZE_F size = pRenderTarget->GetSize();

        const float x = size.width / 2.0F;
        const float y = size.height / 2.0F;
        const float radius = (((x) < (y)) ? (x) : (y));

        ellipseA = D2D1::Ellipse(D2D1::Point2F(x, y), radius, radius);

        const float halfX = x / 2.0F;
        const float halfY = y / 2.0F;
        const float halfPi = radius / 2.0F;

        ellipseB = D2D1::Ellipse(D2D1::Point2F(halfX, halfY), halfPi, halfPi);

        const float quartX = halfX / 2.0F;
        const float quartY = halfY / 2.0F;
        const float quartPi = halfPi / 2.0F;

        ellipseC = D2D1::Ellipse(D2D1::Point2F(quartX, quartY), quartPi, quartPi);
    }
}

/**
 * @brief
 *
 */
void MainComponent::OnPaint()
{
    HRESULT hr = CreateGraphicsResources();
    if (SUCCEEDED(hr))
    {
        PAINTSTRUCT ps;
        BeginPaint(m_hwnd, &ps);

        pRenderTarget->BeginDraw();

        pRenderTarget->Clear(D2D1::ColorF(D2D1::ColorF::CadetBlue));
        pRenderTarget->FillEllipse(ellipseA, pBrush);
        pRenderTarget->FillEllipse(ellipseB, pBrushB);
        pRenderTarget->FillEllipse(ellipseC, pBrushC);

        hr = pRenderTarget->EndDraw();
        if (FAILED(hr) || hr == D2DERR_RECREATE_TARGET)
        {
            DiscardGraphicsResources();
        }
        EndPaint(m_hwnd, &ps);
    }
}

/**
 * @brief
 *
 * @return HRESULT
 */
HRESULT MainComponent::CreateGraphicsResources()
{
    HRESULT hr = S_OK;

    if (pRenderTarget == NULL)
    {
        RECT rc;
        GetClientRect(m_hwnd, &rc);

        D2D1_SIZE_U size = D2D1::SizeU(rc.right, rc.bottom);

        hr = pFactory->CreateHwndRenderTarget(
            D2D1::RenderTargetProperties(),
            D2D1::HwndRenderTargetProperties(m_hwnd, size),
            &pRenderTarget
        );

        if (SUCCEEDED(hr))
        {
            const D2D1_COLOR_F color = D2D1::ColorF(1.0F, 1.0F, 0.0F, 0.5F);
            hr = pRenderTarget->CreateSolidColorBrush(color, &pBrush);

            if (SUCCEEDED(hr))
            {
                const D2D1_COLOR_F colorB = D2D1::ColorF(1.0F, 0.0F, 0.0F, 0.5F);
                hr = pRenderTarget->CreateSolidColorBrush(colorB, &pBrushB);

                if (SUCCEEDED(hr))
                {
                    const D2D1_COLOR_F colorC = D2D1::ColorF(0.0F, 0.0F, 1.0F, 0.5F);
                    hr = pRenderTarget->CreateSolidColorBrush(colorC, &pBrushC);

                    if (SUCCEEDED(hr))
                    {
                        CalculateLayout();
                    }
                }
            }
        }
    }
    return hr;
}

/**
 * @brief
 *
 */
void MainComponent::DiscardGraphicsResources()
{
    SafeRelease(&pRenderTarget);
    SafeRelease(&pBrush);
    SafeRelease(&pBrushB);
    SafeRelease(&pBrushC);
}

/**
 * @brief
 *
 * @return HRESULT
 */
HRESULT MainComponent::OnCreate()
{
    return (D2D1CreateFactory(D2D1_FACTORY_TYPE_SINGLE_THREADED, &pFactory));
}

/**
 * @brief
 *
 */
void MainComponent::OnClose() CONST
{
    if (MessageBox(m_hwnd, TEXT("Sure?"), ClassName(), MB_OKCANCEL) == IDOK)
    {
        DestroyWindow(m_hwnd);
    }
}

/**
 * @brief
 *
 */
void MainComponent::OnDestroy()
{
    DiscardGraphicsResources();
    SafeRelease(&pFactory);
    PostQuitMessage(0);
}
